ogr2ogr with MySQL 8: POINT column doesn't have SRID attribute

I imported shape file using ogr2ogr and I checked created table. Created table doesn't have SIRD attribute on POINT column.

## Expected behavior and actual behavior.
Created table have SIRD attribute on POINT column.

Note: 
POINT column of MySQL 8.0 can have SRID attribute. I think adding SRID attribute is desirable to avoid mistakes.

MySQL 8.0 Reference Manual / 11.5.1 Spatial Data Types
https://dev.mysql.com/doc/refman/8.0/en/spatial-type-overview.html
=> Columns with a spatial data type can have an SRID attribute, to explicitly indicate the spatial reference system (SRS) for values stored in the column.

## Steps to reproduce the problem.
1. Import shape file using ogr2ogr
ex) C:\> ogr2ogr -f "MySQL" MYSQL:geotest,user=root,password=root C:\work\point_2449.shp

2. Execute "show create table" for created table
ex) mysql> show create table point_2449\G
*************************** 1. row ***************************
       Table: point_2449
Create Table: CREATE TABLE `point_2449` (
  `OGR_FID` int(11) NOT NULL AUTO_INCREMENT,
  `SHAPE` geometry NOT NULL,
  `gid` decimal(10,0) DEFAULT NULL,
  `point_name` varchar(254) DEFAULT NULL,
  UNIQUE KEY `OGR_FID` (`OGR_FID`),
  SPATIAL KEY `SHAPE` (`SHAPE`)
) ENGINE=InnoDB AUTO_INCREMENT=4 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
1 row in set (0.00 sec)

* I hope modify CREATE TABLE as follows.
Before:  `SHAPE` geometry NOT NULL,
After:  `SHAPE` geometry NOT NULL SRID 2449,

## Operating system
Windows 10 x64

## GDAL version and provenance
C:\> ogr2ogr --version
GDAL 2.3.2, released 2018/09/21

@YoshiakiYamasaki What is a correct behavior for  GDAL/ogr when accepting such SRID value?

   Simply adding SRID block brings  autotest case `test_ogr_mysql_longlat` failure.

```
     def test_ogr_mysql_longlat():
    
        if gdaltest.mysql_ds is None:
            pytest.skip()
    
        srs = osr.SpatialReference()
        srs.SetFromUserInput('WGS84')
        lyr = gdaltest.mysql_ds.CreateLayer('ogr_mysql_longlat',
                                            geom_type=ogr.wkbPoint,
                                            srs=srs,
                                            options=['ENGINE=MyISAM'])
>       f = ogr.Feature(lyr.GetLayerDefn())
E       AttributeError: 'NoneType' object has no attribute 'GetLayerDefn'

ogr_mysql.py:880: AttributeError
-------------------------------------------- Captured stderr call ---------------------------------------------
MYSQL: Creating layer ogr_mysql_longlat.
MYSQL: Geometry Column Name SHAPE.
MYSQL: FID Column Name OGR_FID.
ERROR 1: MySQL error message:The storage engine for the table doesn't support geographic spatial reference systems Description: CREATE TABLE `ogr_mysql_longlat` (    OGR_FID INT UNIQUE NOT NULL AUTO_INCREMENT,    SHAPE GEOMETRY NOT NULL /*!80003 SRID 4326 */) ENGINE = MyISAM
```

Without SRID, before adding SRID, it passes.
Is it expected behavior for MySQL users? Should the case  fail because of mis-configured combination of ENGINE and SRID?

--- a/autotest/ogr/ogr_mysql.py
+++ b/autotest/ogr/ogr_mysql.py
@@ -91,7 +91,7 @@ def test_ogr_mysql_2():
     ######################################################
     # Create Layer
     gdaltest.mysql_lyr = gdaltest.mysql_ds.CreateLayer('tpoly', srs=shp_lyr.GetSpatialRef(),
-                                                       options=['ENGINE=MyISAM'])
+                                                       options=[])
 
     ######################################################
     # Setup Schema
@@ -543,7 +543,7 @@ def test_ogr_mysql_20():
     if gdaltest.mysql_ds is None:
         pytest.skip()
 
-    layer = gdaltest.mysql_ds.CreateLayer('select', options=['ENGINE=MyISAM'])
+    layer = gdaltest.mysql_ds.CreateLayer('select', options=[])
     ogrtest.quick_create_layer_def(layer,
                                    [('desc', ogr.OFTString),
                                     ('select', ogr.OFTString)])
@@ -572,7 +572,7 @@ def test_ogr_mysql_21():
     if gdaltest.mysql_ds is None:
         pytest.skip()
 
-    layer = gdaltest.mysql_ds.CreateLayer('tablewithspatialindex', geom_type=ogr.wkbPoint, options=['ENGINE=MyISAM'])
+    layer = gdaltest.mysql_ds.CreateLayer('tablewithspatialindex', geom_type=ogr.wkbPoint, options=[])
     ogrtest.quick_create_layer_def(layer, [('name', ogr.OFTString)])
     dst_feat = ogr.Feature(feature_def=layer.GetLayerDefn())
     dst_feat.SetField('name', 'name')
@@ -598,7 +598,7 @@ def test_ogr_mysql_22():
         pytest.skip()
 
     layer = gdaltest.mysql_ds.CreateLayer('tablewithoutspatialindex', geom_type=ogr.wkbPoint,
-                                          options=['SPATIAL_INDEX=NO', 'ENGINE=MyISAM'])
+                                          options=['SPATIAL_INDEX=NO'])
     ogrtest.quick_create_layer_def(layer, [('name', ogr.OFTString)])
     dst_feat = ogr.Feature(feature_def=layer.GetLayerDefn())
     dst_feat.SetField('name', 'name')
@@ -713,7 +713,7 @@ def test_ogr_mysql_25():
     if gdaltest.mysql_ds is None:
         pytest.skip()
 
-    lyr = gdaltest.mysql_ds.CreateLayer('ogr_mysql_25', geom_type=ogr.wkbPoint, options=['ENGINE=MyISAM'])
+    lyr = gdaltest.mysql_ds.CreateLayer('ogr_mysql_25', geom_type=ogr.wkbPoint, options=[])
     field_defn = ogr.FieldDefn('field_not_nullable', ogr.OFTString)
     field_defn.SetNullable(0)
     lyr.CreateField(field_defn)
@@ -762,7 +762,7 @@ def test_ogr_mysql_26():
     if gdaltest.mysql_ds is None:
         pytest.skip()
 
-    lyr = gdaltest.mysql_ds.CreateLayer('ogr_mysql_26', geom_type=ogr.wkbPoint, options=['ENGINE=MyISAM'])
+    lyr = gdaltest.mysql_ds.CreateLayer('ogr_mysql_26', geom_type=ogr.wkbPoint, options=[])
 
     field_defn = ogr.FieldDefn('field_string', ogr.OFTString)
     field_defn.SetDefault("'a''b'")
@@ -833,6 +833,36 @@ def test_ogr_mysql_26():
     gdal.Unlink('/vsimem/ogr_gpkg_24.gpkg')
 
 ###############################################################################
+# Test created table indecs
+
+def test_ogr_mysql_27():
+
+    if gdaltest.mysql_ds is None:
+        pytest.skip()
+
+    if not gdaltest.is_mysql_8_or_later:
+        pytest.skip()
+
+    layer = gdaltest.mysql_ds.GetLayerByName('tpoly')
+    if layer is None:
+        pytest.skip('did not get tpoly layer')
+
+    sql_lyr = gdaltest.mysql_ds.ExecuteSQL('SHOW CREATE TABLE tpoly')
+    f = sql_lyr.GetNextFeature()
+    field = f.GetField(1)
+    res = False
+    for line in field.splitlines():
+        if 'geometry' in line:
+            if "SRID" in line:
+                res = True
+            else:
+                res = False
+    if not res:
+        print('{}'.format(field))
+        pytest.fail("Not found SRID definition with GEOMETORY field.")
+    gdaltest.mysql_ds.ReleaseResultSet(sql_lyr)
+
+###############################################################################
 #
 
 
@@ -842,11 +872,11 @@ def test_ogr_mysql_longlat():
         pytest.skip()
 
     srs = osr.SpatialReference()
-    srs.SetFromUserInput('WGS84')
+    srs.SetFromUserInput('EPSG:4326')
     lyr = gdaltest.mysql_ds.CreateLayer('ogr_mysql_longlat',
                                         geom_type=ogr.wkbPoint,
                                         srs=srs,
-                                        options=['ENGINE=MyISAM'])
+                                        options=[])
     f = ogr.Feature(lyr.GetLayerDefn())
     geom = ogr.CreateGeometryFromWkt('POINT(150 2)')
     f.SetGeometry(geom)
@@ -868,6 +898,21 @@ def test_ogr_mysql_longlat():
         print(extent)
         pytest.fail('Extents do not match')
 
+    if gdaltest.is_mysql_8_or_later:
+        sql_lyr = gdaltest.mysql_ds.ExecuteSQL('SHOW CREATE TABLE ogr_mysql_longlat')
+        f = sql_lyr.GetNextFeature()
+        field = f.GetField(1)
+        res = False
+        for line in field.splitlines():
+            if 'geometry' in line:
+                if "SRID" in line:
+                    res = True
+                else:
+                    res = False
+        if not res:
+            print('{}'.format(field))
+            pytest.fail("Not found SRID definition with GEOMETORY field.")
+        gdaltest.mysql_ds.ReleaseResultSet(sql_lyr)
 
     
 ###############################################################################

--- a/gdal/ci/travis/ubuntu_1804/before_install.sh
+++ b/gdal/ci/travis/ubuntu_1804/before_install.sh
@@ -14,7 +14,7 @@ sleep 10
 docker exec -it sql1 /opt/mssql-tools/bin/sqlcmd -l 30 -S localhost -U SA -P DummyPassw0rd -Q "CREATE DATABASE TestDB;"
 
 # MySQL 8
-docker run --name mysql1 -e MYSQL_ROOT_PASSWORD=passwd -e "MYSQL_ROOT_HOST=%" -p 33060:3306 -d mysql/mysql-server:8.0.1 mysqld --default-authentication-plugin=mysql_native_password
+docker run --name mysql1 -e MYSQL_ROOT_PASSWORD=passwd -e "MYSQL_ROOT_HOST=%" -p 33060:3306 -d mysql/mysql-server:8.0.18 mysqld --default-authentication-plugin=mysql_native_password
 
 # MariaDB 10.3.9
 docker run --name mariadb -e MYSQL_ROOT_PASSWORD=passwd -e "MYSQL_ROOT_HOST=%" -p 33061:3306 -d mariadb:10.3.9

--- a/gdal/ogr/ogrsf_frmts/mysql/ogrmysqldatasource.cpp
+++ b/gdal/ogr/ogrsf_frmts/mysql/ogrmysqldatasource.cpp
@@ -1032,6 +1032,7 @@ OGRMySQLDataSource::ICreateLayer( const char * pszLayerNameIn,
 
     const char *pszSI = CSLFetchNameValue( papszOptions, "SPATIAL_INDEX" );
     const bool bHasSI = ( eType != wkbNone && (pszSI == nullptr || CPLTestBool(pszSI)) );
+    int nSRSId = GetUnknownSRID();
 
     if( wkbFlatten(eType) == wkbNone )
     {
@@ -1042,12 +1043,22 @@ OGRMySQLDataSource::ICreateLayer( const char * pszLayerNameIn,
     }
     else
     {
-        osCommand.Printf(
-                 "CREATE TABLE `%s` ( "
-                 "   %s %s UNIQUE NOT NULL AUTO_INCREMENT, "
-                 "   %s GEOMETRY %s)",
-                 pszLayerName, pszExpectedFIDName, pszFIDType, pszGeomColumnName,
-                 bHasSI ? "NOT NULL" : "");
+        // when using mysql8 and SRS is specified, use SRID option for geometry.
+        if( GetMajorVersion() < 8 || IsMariaDB() || poSRS == nullptr ||
+            (nSRSId = FetchSRSId( poSRS )) == GetUnknownSRID() )
+            osCommand.Printf(
+                    "CREATE TABLE `%s` ( "
+                    "   %s %s UNIQUE NOT NULL AUTO_INCREMENT, "
+                    "   %s GEOMETRY %s)",
+                    pszLayerName, pszExpectedFIDName, pszFIDType, pszGeomColumnName,
+                    bHasSI ? "NOT NULL" : "");
+        else
+            osCommand.Printf(
+                    "CREATE TABLE `%s` ( "
+                    "   %s %s UNIQUE NOT NULL AUTO_INCREMENT, "
+                    "   %s GEOMETRY %s /*!80003 SRID %d */)",
+                    pszLayerName, pszExpectedFIDName, pszFIDType, pszGeomColumnName,
+                    bHasSI ? "NOT NULL" : "", nSRSId);
     }
 
     if( CSLFetchNameValue( papszOptions, "ENGINE" ) != nullptr )
@@ -1083,12 +1094,14 @@ OGRMySQLDataSource::ICreateLayer( const char * pszLayerNameIn,
 
 /* -------------------------------------------------------------------- */
 /*      Try to get the SRS Id of this spatial reference system,         */
-/*      adding tot the srs table if needed.                             */
+/*      adding to the srs table if needed.                             */
 /* -------------------------------------------------------------------- */
-    int nSRSId = GetUnknownSRID();
-
-    if( poSRS != nullptr )
-        nSRSId = FetchSRSId( poSRS );
+    if( GetMajorVersion() < 8 || IsMariaDB() )
+    {
+        nSRSId = GetUnknownSRID();
+        if (poSRS != nullptr)
+            nSRSId = FetchSRSId(poSRS);
+    }
 
 /* -------------------------------------------------------------------- */
 /*      Sometimes there is an old crufty entry in the geometry_columns  */

--- a/gdal/ogr/ogrsf_frmts/mysql/ogrmysqldatasource.cpp
+++ b/gdal/ogr/ogrsf_frmts/mysql/ogrmysqldatasource.cpp
@@ -1032,7 +1032,18 @@ OGRMySQLDataSource::ICreateLayer( const char * pszLayerNameIn,
 
     const char *pszSI = CSLFetchNameValue( papszOptions, "SPATIAL_INDEX" );
     const bool bHasSI = ( eType != wkbNone && (pszSI == nullptr || CPLTestBool(pszSI)) );
+
+    // Calling this does no harm
+    InitializeMetadataTables();
+
+/* -------------------------------------------------------------------- */
+/*      Try to get the SRS Id of this spatial reference system,         */
+/*      adding to the srs table if needed.                             */
+/* -------------------------------------------------------------------- */
+
     int nSRSId = GetUnknownSRID();
+    if (poSRS != nullptr)
+        nSRSId = FetchSRSId(poSRS);
 
     if( wkbFlatten(eType) == wkbNone )
     {
@@ -1044,8 +1055,7 @@ OGRMySQLDataSource::ICreateLayer( const char * pszLayerNameIn,
     else
     {
         // when using mysql8 and SRS is specified, use SRID option for geometry.
-        if( GetMajorVersion() < 8 || IsMariaDB() || poSRS == nullptr ||
-            (nSRSId = FetchSRSId( poSRS )) == GetUnknownSRID() )
+        if( GetMajorVersion() < 8 || IsMariaDB() || nSRSId == GetUnknownSRID() )
             osCommand.Printf(
                     "CREATE TABLE `%s` ( "
                     "   %s %s UNIQUE NOT NULL AUTO_INCREMENT, "
@@ -1088,19 +1098,6 @@ OGRMySQLDataSource::ICreateLayer( const char * pszLayerNameIn,
     if( hResult != nullptr )
         mysql_free_result( hResult );
     hResult = nullptr;
-
-    // Calling this does no harm
-    InitializeMetadataTables();
-
-/* -------------------------------------------------------------------- */
-/*      Try to get the SRS Id of this spatial reference system,         */
-/*      adding to the srs table if needed.                             */
-/* -------------------------------------------------------------------- */
-    if( GetMajorVersion() < 8 || IsMariaDB() )
-    {
-        if (poSRS != nullptr)
-            nSRSId = FetchSRSId(poSRS);
-    }
 
 /* -------------------------------------------------------------------- */
 /*      Sometimes there is an old crufty entry in the geometry_columns  */

--- a/gdal/ogr/ogrsf_frmts/mysql/ogrmysqldatasource.cpp
+++ b/gdal/ogr/ogrsf_frmts/mysql/ogrmysqldatasource.cpp
@@ -1098,7 +1098,6 @@ OGRMySQLDataSource::ICreateLayer( const char * pszLayerNameIn,
 /* -------------------------------------------------------------------- */
     if( GetMajorVersion() < 8 || IsMariaDB() )
     {
-        nSRSId = GetUnknownSRID();
         if (poSRS != nullptr)
             nSRSId = FetchSRSId(poSRS);
     }

